### ISSUE TITLE
Cannot delete DAG with many runs

### ISSUE THREAD

**Issue by GergelyKalmar** [apache/airflow#41148]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.8.1 (AWS MWAA)

### What happened?

I tried to delete a DAG with many runs (14000+). After about 4 minutes I receive "xxx.yyy.airflow.amazonaws.com didn’t send any data. ERR_EMPTY_RESPONSE", and the DAG is not deleted.

### What you think should happen instead?

The DAG should have been deleted.

### How to reproduce

Create a DAG with 14000 runs and try to delete it.

### Operating System

AWS MWAA

### Deployment

Amazon (AWS) MWAA

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by eladkal** [comment]
yes this is a bug but just to clarify the delete partially works.
It deletes some of the records before raising the error, so every time you click on the button it will reduce the amount of records. I assume 2-3 times will delete all the records

> and the DAG is not deleted.

DAG can not be deleted from the UI. the delete only removes the records associated with it (DagRun, TaskInstance...)
if you want to remove dag completely then you need also to remove the .py file of the DAG and then delete the records (with the UI or API if the dag is not present)

**Comment by eladkal** [comment]
tagged with DB label as this is probably related to inefficient query to the DB.

**Comment by omkar-foss** [comment]
>I tried to delete a DAG with many runs (14000+). After about 4 minutes I receive "xxx.yyy.airflow.amazonaws.com didn’t send any data. ERR_EMPTY_RESPONSE", and the DAG is not deleted.

@GergelyKalmar the `ERR_EMPTY_RESPONSE` suggests that the connection is closed before the DAG deletion completed, particularly because the  of 14k+ runs deletion operation [in this function](https://github.com/apache/airflow/blob/main/airflow/api/common/delete_dag.py#L42-L108) is taking too long. This is a bug as @eladkal pointed out, and when this bug gets fixed here in upstream Airflow, it may be a while before it reflects in downstream MWAA which you're using.

Meanwhile, I'd suggest you try to delete the DAG via the Airflow REST API with an increased request timeout. I've written a small script for this ([Gist here](https://gist.github.com/omkar-foss/51fe2e79a8da628833412652e69f32d1)) by referring to the MWAA docs, please check it out. You can tweak the script as per your need and increase the timeout beyond 10 mins if the deletion takes longer and fails. Hope this helps.
